File: javawsServer/OneOhOneCompanies/src/org/softlang/service/company/Department.java
package org.softlang.service.company;

import java.io.Serializable;
import java.util.LinkedList;
import java.util.List;

/**
 * A department has a name, a manager, employees, and subdepartments.
 */
public class Department implements Serializable {

	private static final long serialVersionUID = -2008895922177165250L;
	
	private String name;
	private Employee manager;
	private List<Department> subdepts = new LinkedList<Department>();
	private List<Employee> employees = new LinkedList<Employee>();

	public String getName() {
		return name;
	}

	public void setName(String name) {
		this.name = name;
	}

	public Employee getManager() {
		return manager;
	}

	public void setManager(Employee manager) {
		this.manager = manager;
	}

	public List<Department> getSubdepts() {
		return subdepts;
	}

	public List<Employee> getEmployees() {
		return employees;
	}
	
	public double total() {
		double total = 0;
		total += getManager().getSalary();
		for (Department s : getSubdepts())
			total += s.total();
		for (Employee e : getEmployees())
			total += e.getSalary();
		return total;		
	}	
	
	public void cut() {
		getManager().cut();
		for (Department s : getSubdepts())
			s.cut();
		for (Employee e : getEmployees())
			e.cut();
	}	
}

File: qt/QtImplementation/department.cpp
#include "department.h"
#include <iostream>

using namespace std;

Department::Department(){}

Department::Department(QString name)
{
    this->name = name;
}

void Department::addEmployee(Employee e)
{
    this->employees.append(e);
}

void Department::addSubDepartment(Department d)
{
    this->subdepartments.append(d);
}

void Department::setManager(Employee e)
{
    this->manager = e;
}

Employee Department::getManager()
{
    return manager;
}

void Department::setName(QString name)
{
    this->name = name;
}

QString Department::getName()
{
    return name;
}

QVector<Employee> Department::getEmployees()
{
    return employees;
}

QVector<Department> Department::getDepartments()
{
    return subdepartments;
}

void Department::cut()
{
    manager.cut();

    for(QVector<Employee>::iterator it = employees.begin(); it != employees.end(); it++)
    {
        it->cut();
    }

    for(QVector<Department>::iterator it = subdepartments.begin(); it != subdepartments.end(); it++)
    {
        it->cut();
    }
}

double Department::total()
{
    double total = 0.0;

    total += manager.getSalary();

    for(QVector<Employee>::iterator it = employees.begin(); it != employees.end(); it++)
    {
        total += it->getSalary();
    }

    for(QVector<Department>::iterator it = subdepartments.begin(); it != subdepartments.end(); it++)
    {
        total += it->total();
    }
    return total;
}

void Department::walkDepartment(QTreeWidgetItem* parent)
{

    QTreeWidgetItem *current = new QTreeWidgetItem();
    current->setText(0,manager.getName());

    Registry::makeEntry(manager);
    QVariant iD = manager.getID();
    manager.setTypeInfo('e');
    current->setData(0,32,iD);

    parent->addChild(current);

    for(QVector<Employee>::iterator it = employees.begin(); it != employees.end(); it++)
    {
        QTreeWidgetItem *current = new QTreeWidgetItem();
        current->setText(0,it->getName());
        Registry::makeEntry(*it);
        QVariant iD = it->getID();
        it->setTypeInfo('e');
        current->setData(0,32,iD);

        parent->addChild(current);

  }
    for(QVector<Department>::iterator it = subdepartments.begin(); it != subdepartments.end(); it++)
    {
        QTreeWidgetItem *current = new QTreeWidgetItem();
        current->setText(0,it->getName());

        Registry::makeEntry(*it);
        QVariant iD = it->getID();
        it->setTypeInfo('d');
        current->setData(0,32,iD);

        parent->addChild(current);
        it->walkDepartment(current);
    }
}
